build: JavaFX 18 -> 19-ea+11, Kotlin 1.6.0 -> 1.7.10, attach 4.0.13 -> 4.0.15, jackson 2.12.1 -> 2.13.3

chengenzhao opened the pull request:
Hi:

We tried to bump the versions of dependencies and plugins
JavaFX from 18 -> 19-ea+10, it should be fine to change the final version to 19 since the JavaFX 19 is coming to the final release candidate, won't change too much
and JavaFX 19 has an important changes which is adding new .map and .flatMap method which would be really helpful for binding style programming

Kotlin from 1.6.0 -> 1.7.10
We have to remove the modular classifier of dependency of Kotlin-stdlib
since the incremental compilation is not experimental any more
and also another significant change in the code is that
@Suppress("NON_EXHAUSTIVE_WHEN") is not available since 1.7
non exhaustive when statement will cause compilation error
so use else clause to complete the when statement

and also bump the plugins' versions

The bumped version tests have passed in my computer
<img width="1180" alt="截屏2022-08-27 上午3 51 51" src="https://user-images.githubusercontent.com/5525436/186981813-920b402d-1eae-40c2-bf2d-70d7164dbe28.png">


Cheers

chengenzhao committed:
bump plugin and dependencies versions

chengenzhao committed:
bump kotlin version to 1.7.10

chengenzhao committed:
when clause must be exhaustive now
so add else clause and return the compare result of modifier == NONE
and also lift return to the place before when.

chengenzhao committed:
when clause must be exhaustive now
so add else clause and return false result
and also lift return to the place before when.

chengenzhao committed:
@Suppress("NON_EXHAUSTIVE_WHEN")
is not available since 1.7
so use else clause to replace it

chengenzhao commented on fxgl-core/src/main/kotlin/com/almasb/fxgl/input/Triggers.kt:
```diff
@@ -152,13 +152,13 @@ data class MouseTrigger
+                KeyCode.CONTROL -> modifier == InputModifier.CTRL
+                KeyCode.SHIFT -> modifier == InputModifier.SHIFT
+                KeyCode.ALT -> modifier == InputModifier.ALT
+                else -> false
```
Although this code has passed the test
but still not sure about the return value of else
@AlmasB would you mind checking this line of code?
Is it return isTriggered(event) or false?

AlmasB replied:
Originally, `else` would just not run, falling through till the end, so I think it's `return isTriggered(event)`.

chengenzhao replied:
Done

AlmasB commented on fxgl/src/main/kotlin/com/almasb/fxgl/dev/Console.kt:
```diff
@@ -92,10 +92,10 @@ class Console : Pane() {
                 when (it.code) {
                     KeyCode.UP -> prev()
                     KeyCode.DOWN -> next()
+                    else -> Unit
```
Fine with this, the `dev` package needs redesigning anyway

AlmasB commented on pom.xml:
```diff
@@ -75,37 +75,37 @@
-        <maven.gpg.version>1.6</maven.gpg.version>
-        <sonatype.nexus.staging.version>1.6.8</sonatype.nexus.staging.version>
-        <maven.compiler.version>3.8.0</maven.compiler.version>
+        <maven.gpg.version>3.0.1</maven.gpg.version>
```
Could we revert all plugins back to their original versions (unless they are required by the dependencies) so we can separate any potential implications.

chengenzhao replied:
reverted
except sonatype.nexus.staging.version has to be bumped because the kotlin dependencies require higher version of this plugin

AlmasB commented on pom.xml:
```diff
@@ -75,37 +75,37 @@
-        <junit.jupiter.version>5.8.2</junit.jupiter.version>
-        <junit.platform.version>1.8.2</junit.platform.version>
-        <hamcrest.version>1.3</hamcrest.version>
+        <junit.jupiter.version>5.9.0</junit.jupiter.version>
```
Could we revert test deps back too, can handle them in a separate commit.

chengenzhao replied:
reverted

AlmasB commented on pom.xml:
```diff
@@ -75,37 +75,37 @@
             <artifactId>kotlin-stdlib</artifactId>
             <version>${kotlin.version}</version>
-            <classifier>modular</classifier>
+<!--            <classifier>modular</classifier>-->
```
Given there is no classifier now, we can delete this line.

chengenzhao replied:
deleted

AlmasB reviewed (commented):
Thanks for this, please can you have a look at the comments. Meanwhile, I'll check the logic behind `isReleased()` and get back.

chengenzhao committed:
revert the test and unnecessary plugins' version back
and also delete classifier of kotlin dependency

chengenzhao commented:
> Thanks for this, please can you have a look at the comments. Meanwhile, I'll check the logic behind `isReleased()` and get back.

Sure, un necessary bumped plugins and test dependencies are reverted

AlmasB commented:
Looks good, ready for merge once `isReleased()` is updated.

chengenzhao commented on fxgl-core/src/main/kotlin/com/almasb/fxgl/input/Triggers.kt:
```diff
@@ -152,13 +152,13 @@ data class MouseTrigger
+                KeyCode.CONTROL -> modifier == InputModifier.CTRL
+                KeyCode.SHIFT -> modifier == InputModifier.SHIFT
+                KeyCode.ALT -> modifier == InputModifier.ALT
+                else -> false
```
```suggestion
                else -> isTriggered(event)
```

chengenzhao committed:
Update fxgl-core/src/main/kotlin/com/almasb/fxgl/input/Triggers.kt

chengenzhao committed:
plugin jacoco requires new version of 0.8.8
because 0.8.7 doesn't support class file version 62 which is JDK 18
I tried to use jacoco 0.8.7 it throws out some warning exceptions telling me:Unsupported class file major version 62

So I think it would be better to upgrade jacoco plugin also

chengenzhao commented:
> Looks good, ready for merge once `isReleased()` is updated.

isReleased() is updated

but we got another problem which may be caused by jacoco plugin version 0.8.7
seems like jacoco 0.8.7 doesn't support class file format class file major version 62 which is Java 18 class file
It may cause the test of fxgl-io failed if maven uses JDK 18 to build the FXGL projects
So should we upgrade jacoco plugin version also?
Bumping the jacoco plugin version to 0.8.8 and the CI/CD passed all tests

Cheers

chengenzhao committed:
bump jfx version to 19-ea+11

AlmasB reviewed (approved):
looks good, many thanks
